include/LightGBM/dataset.h

```diff
@@ -834,6 +938,7 @@ class Dataset {
   bool zero_as_missing_;
   std::vector<int> feature_need_push_zeros_;
   std::vector<std::vector<float>> raw_data_;
+  bool wait_for_manual_finish_;
```

Comment:
should we use a mutex when setting wait_for_manual_finish_?

Reply:
Sure, to be absolutely safe. Note that dataset does not currently have a mutex, but I added one.